Ally_Desktop_Assistant.py

```diff
@@ -0,0 +1,464 @@
+from tkinter import *
+import tkinter.messagebox as tmsg
+from tkinter.filedialog import askdirectory
+import pyttsx3
+import speech_recognition as sr
+import time
+import webbrowser
+import json
+import os
+import random
+
+name = "Ally"
+name = name.title()
+
+data_filename = "Ally_Data.json"
+history_filename = "Ally_History.txt"
+tut_address = r"https://www.youtube.com/watch?v=v6YqIklzYZU"
+
+
+def show(data):
+    time_to_show = time.strftime("(%I:%M %p)")
+    chat_history.configure(state=NORMAL)
+    chat_history.insert(END, f"{time_to_show} {data}\n")
+    chat_history.yview_pickplace(END)
+    chat_history.update()
+    chat_history.configure(state=DISABLED)
+    with open(history_filename, "a") as f_obj:
+        f_obj.write(f"{time.ctime()}  {data}\n")
+
+
+def time_am_pm():
+
+    pr_time = time.strftime('%I:%M %p')
+    return pr_time
+
+
+def get_date():
+    mon_dict = {1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June', 7: 'July', 8: 'August',
+                9: 'September', 10: 'October', 11: 'November', 12: 'December'}
+    date_day = str(int(time.strftime("%d")))
+    date_mon_num = int(time.strftime("%m"))
+    date_mon = mon_dict[date_mon_num]
+    date_year = str(int(time.strftime("%Y")))
+    date_week_day = (time.strftime("%A"))
+    today_date = f"{date_week_day} the {date_day}th of {date_mon} {date_year}"
+    return today_date
+
+
+def show_status(string):
+    status_var.set(string)
+    status_label.update()
+
+
+def change_voice_code(event):
+    global temp_voice_code
+
+    voice_type = (voice_var.get())
+    filename = data_filename
+    try:
+        with open(filename, "r") as f_obj:
+            data_dict = json.load(f_obj)
+            data_dict["voice"] = voice_type
+
+        with open(filename, 'w') as file_object:
```

Comment:
Why we are open the same file with r and w operation rather than opening it as read+write(`‘r+’`) in one go